lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -57,140 +59,124 @@ def tags(repository, params = {})
           unauthenticated_repos
         else
-          (unauthenticated_repos + user.repositories_dataset.map(:name)).sort
+          # TODO implement as a single query
+          unauthenticated_repos | user.repositories_dataset.select(:name)
         end
       end
 
       def unauthenticated_repos
-        Repository.where(auth_required: false).order(:name).map(:name)
+        database[:repository].where(auth_required: false).order(:name).select(:name)
       end
 
       # Replaces the entire list of repositories
       def update_repository_list(repo_list)
-        RepositoryUser.dataset.delete
-        Repository.dataset.delete
-        repo_list.each do |repo|
-          Repository.find_or_create(name: repo['repository'],
-                                    auth_required: repo['auth_required'].to_s.downcase == "true")
-        end
+        # TODO: make repositories_users cascade on delete repository
+        database[:repositories_users].delete
+
+        repository = database[:repository]
+        repository.delete
+
+        repository.import(
+          [:name, :auth_required],
+          repo_list.map { |repo| [repo['repository'], repo['auth_required'].to_s.downcase == "true"] },
+        )
       end
 
       # Replaces the entire user-repo mapping for all logged-in users
       def update_user_repo_mapping(user_repo_maps)
         # Get hash map of all users and their repositories
         # Ex: {"users"=> [{"admin"=>[{"repository"=>"repo", "auth_required"=>"true"}]}]}
         # Go through list of repositories and add them to the DB
-        RepositoryUser.dataset.delete
-        user_repo_maps['users'].each do |user_repo_map|
-          user_repo_map.each do |user, repos|
-            next if repos.nil?
-
-            repos.each do |repo|
-              found_repo = Repository.find(name: repo['repository'],
-                                           auth_required: repo['auth_required'].to_s.downcase == "true")
-              if found_repo.nil?
-                logger.warn("#{repo['repository']} does not exist in this smart proxy's environments")
-              elsif found_repo.auth_required
-                found_repo.add_user(User.find(name: user))
-              end
-            end
+        repositories = database[:repositories]
+
+        entries = user_repo_maps['users'].flat_map do |user_repo_map|
+          user_repo_map.filter_map do |username, repos|
+            user_repo_names = repos.filter { |repo| repo['auth_required'].to_s.downcase == "true" }.map { |repo| repo['repository'] }
+            user = database[:user].find(name: username)
+            # TODO: check you can select a static column like this
+            repositories.where(name: user_repo_names, auth_required: true).select(:id, user_id: user.id)
           end
         end
+
+        repository_users = database[:repository_users]
+        repository_users.delete
+        repositories_users.import([:repository_id, :user_id], entries)
```

Comment:
Issues can pop up still during concurrent smart proxy syncs:

```
2024-03-06T16:20:27 848a5696 [W] Error processing request '848a5696-6ca6-4138-898c-36377e3bfbea: <Sequel::UniqueConstraintViolation>: SQLite3::ConstraintException: UNIQUE constraint failed: repositories_users.repository_id, repositories_users.user_id
```

Also, I had to switch to importing DB IDs in `entries` -- seems the Sequel Query there is disagreeing with `import`.  I wonder if it's because the query above `repositories.where(name: user_repo_names, auth_required: true).select(:id, user_id: user.id)` needs a joins to `repositories_users`.